File: src/python/algorithms/misc/sqrt.py
def sqrt1(num):
    if num < 2:
        return num
    left, right = 1, num // 2
    result = 0
    while left <= right:
        mid = left + (right - left) // 2
        if mid <= num // mid:
            result = mid
            left = mid + 1
        else:
            right = mid - 1
    return result

def sqrt2(num):
    if num < 2:
        return num
    result = 0
    for i in range(1, num // 2 + 1):
        if i * i > num:
            return result
        result = i
    return result

if __name__ == "__main__":
    test_cases = [0, 1, 4, 8, 9, 15, 16, 25, 26, 100, 101]
    print("Testing sqrt1:")
    for n in test_cases:
        print(f"sqrt1({n}) = {sqrt1(n)}")
    print("\nTesting sqrt2:")
    for n in test_cases:
        print(f"sqrt2({n}) = {sqrt2(n)}")

File: src/javascript/algorithms/misc/sqrt.js
function sqrt1(num) {
    if (num < 2) return num;
    let left = 1, right = Math.floor(num / 2);
    let result = 0;
    while (left <= right) {
        let mid = Math.floor(left + (right - left) / 2);
        if (mid <= Math.floor(num / mid)) {
            result = mid;
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return result;
}

function sqrt2(num) {
    if (num < 2) return num;
    let result = 0;
    for (let i = 1; i <= Math.floor(num / 2); i++) {
        if (i * i > num) return result;
        result = i;
    }
    return result;
}

if (require.main === module) {
    const testCases = [0, 1, 4, 8, 9, 15, 16, 25, 26, 100, 101];
    console.log("Testing sqrt1:");
    for (const n of testCases) {
        console.log(`sqrt1(${n}) = ${sqrt1(n)}`);
    }
    console.log("\nTesting sqrt2:");
    for (const n of testCases) {
        console.log(`sqrt2(${n}) = ${sqrt2(n)}`);
    }
}
